File: 206_Reverse_Linked_List/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        /*
            Input: head of a singly linked list
            Output: head of reversed linked list

            Strategy
            1.) Array to store the position and nodes of the linked list then reverse it O(2n) time, O(n) space

                std::vector <ListNode*>  position;
                ListNode* curr = head;
            
                while (curr){
                    position.push_back(curr);
                    curr = curr->next;
                }
                for (int i = position.size()-1; i >0; i--){
                    position[i]->next = position[i-1];
                }
                if (!position.empty()){
                    position[0]->next = nullptr;
                    return position[position.size()-1];
                }
                return nullptr;


            2.) Reverse inplace
                ListNode* curr = head;
                ListNode* prev = nullptr;
                ListNode* temp = nullptr;
                while (curr) {
                    temp = curr->next;
                    curr->next = prev;
                    prev = curr;
                    curr = temp;
                }
                return prev;


        */
        ListNode* curr = head;
        ListNode* prev = nullptr;
        ListNode* temp = nullptr;
        while (curr) {
            temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }
        return prev;
        
        
    }
};

File: 206_Reverse_Linked_List/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        /*
            Input: head of linkedlist
            output: head of reversed linked list

            Strategy
            1.) Array and reverse: TC: O(2n), SC: O(n)
                - Loop original linked list once, storing each node sequentially in an array
                - Second pass using the array from the end, link them up
            
            2.) Reverse in place: TC: O(n), SC:O(1)
                - One pointer pointer at the next node
                - one pointer pointing at previous
                - one pointer pointing at current node
                - Upon advancing a new node
                    - next points to next node of the current node
                    - current.next points to previous
                    - previous assign to current
                    - current assign to next node
        */
        ListNode curr = head;
        ListNode next;
        ListNode prev = null;

        while(curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
}
